skillmap/src/App.tsx

```diff
@@ -211,13 +211,79 @@ class AppImpl extends React.Component<AppProps, AppState> {
+                // over the local progress that doesn't exist in the signed in
+                // user
+                const urls = Object.keys(state.user.mapProgress);
+                const newUser: UserState = {
+                    ...state.user,
+                    mapProgress: {}
+                }
+
+                for (const url of urls) {
+                    newUser.mapProgress[url] = {
+                        ...state.user.mapProgress[url],
+                    };
+
+                    if (!localUser.mapProgress[url]) continue;
+
+                    const maps = Object.keys(localUser.mapProgress[url]);
+                    for (const map of maps) {
+                        newUser.mapProgress[url][map] = {
+                            ...state.user.mapProgress[url][map]
+                        };
+
+                        // Only copy over state if the user hasn't started this map yet
+                        if (Object.keys(newUser.mapProgress[url][map].activityState).length !== 0) {
+                            continue;
+                        }
+
+                        const activityState: {[index: string]: ActivityState} = {};
+                        newUser.mapProgress[url][map].activityState = activityState;
+
+                        const signedInProgress = state.user.mapProgress[url][map].activityState;
+                        const localProgress = localUser.mapProgress[url][map].activityState
+
+                        for (const activity of Object.keys(signedInProgress)) {
+                            const oldState = signedInProgress[activity];
+                            activityState[activity] = {
+                                ...oldState,
+                                headerId: oldState.headerId ? (headerMap[oldState.headerId] || oldState.headerId) : oldState.headerId
+                            };
+                        }
+
+                        for (const activity of Object.keys(localProgress)) {
+                            const signedInActivity = signedInProgress[activity];
+                            const localActivity = localProgress[activity];
+                            if ((!signedInActivity || !signedInActivity.headerId) && localActivity.headerId) {
+                                const base = signedInActivity || localActivity;
+                                activityState[activity] = {
+                                    ...base,
+                                    headerId: localActivity.headerId ? (headerMap[localActivity.headerId] || localActivity.headerId) : localActivity.headerId
+                                };
+                            }
+                        }
+                    }
+                }
+
+                this.props.dispatchSetUser(newUser)
+                await saveUserStateAsync(newUser);
+            }
+
             // Tell the editor to send us the cloud status of our projects.
             await this.sendMessageAsync({
```

Comment:
I think with the headerId changes, we won't fetch the status of our....already synced projects ever? I think this should request the status of all headerIds, not just the new ones?

Reply:
true! fixed